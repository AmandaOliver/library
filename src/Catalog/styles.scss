.catalog {
    display: flex;
    flex-flow: wrap row;
    justify-content: space-around;
    position: relative;
    top:80px;
}

.spaceHolder {
    width: 100%;
    clear: both;
    overflow: hidden;
}

.bookWrapper {
    width:100%;
    height:100%;
    position: relative;
}
