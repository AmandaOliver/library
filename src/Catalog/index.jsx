import React, {PureComponent} from 'react'
import Book from '../Book'
import { Consumer } from '../context'
import config from '../configuration'
import { List, AutoSizer } from 'react-virtualized'
import {worker, isLibraryInitialized} from '../libraryLoader'
import './styles.scss'

class Catalog extends PureComponent {
  constructor() {
    super()
    this.state = {
      bookIds:[],
      isLoading: true,
      indexInit: 0,
    }
    this.bindedInitHandler = this.initHandler.bind(this)
  }

  componentDidMount() {
    // listen to the web worker to check when to render the first elements
    worker.addEventListener('message', this.bindedInitHandler)
  }

  initHandler(event) {
    if (isLibraryInitialized()) {
      worker.removeEventListener('message', this.bindedInitHandler)
      this.setState({ isLoading: false })
      this.getMoreBooks()
    }
  }

  getMoreBooks() {
    const { bookIds, indexInit } = this.state
    this.setState({
      bookIds: [
        ...bookIds,
        ...new Array(config.booksPerPage).fill().map((_, i) => i+indexInit)
      ],
      indexInit: indexInit + config.booksPerPage,
    })
  }

  render() {
    return (
      <Consumer>
        {({ isFiltered, getBookById, getLibrarySize }) => {
          const rowRenderer = ({ key, index, style }) => {
            const book = getBookById(index)
            return book && <div key={key} style={style}><Book style={style} book={book} /></div>
          }
          return (
            <div className='catalog'>
              <AutoSizer>
                {({ height, width }) => (
                  <List
                    width={width}
                    height={height}
                    rowCount={getLibrarySize()}
                    rowHeight={config.rowHeight}
                    rowRenderer={rowRenderer.bind(this)}
                  />
                )}
              </AutoSizer>
            </div>
          )
        }}
      </Consumer>

    )
  }
}

export default Catalog
