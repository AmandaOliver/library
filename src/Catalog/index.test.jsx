import React from 'react'
import renderer from 'react-test-renderer';
import Catalog from '.'

const booksArray = [
    {
        name: 'mockName',
        genre: 'mockGenre',
        publishDate: new Date(),
        authorId: 0
    },
    {
        name: 'mockName1',
        genre: 'mockGenre1',
        publishDate: Date.now(),
        authorId: 1,
    }
]

const authorsMap = new Map([
    [0, {
        name: 'mockName',
        gender: 'mockGenre'
    }],
    [1, {
        name: 'mockName1',
        gender: 'mockGenre1'
    }]
])

describe('Catalog', () => {
  it('renders', () => {
    const component = renderer.create(
        <Catalog booksArray={booksArray} authorsMap={authorsMap}/>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
})
