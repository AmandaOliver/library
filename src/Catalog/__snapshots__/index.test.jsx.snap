// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog renders 1`] = `
<div
  className="catalog"
>
  <div
    className="book"
  >
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Name: 
      </span>
      <span
        className="book__property__value"
      >
        mockName
         
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Genre: 
      </span>
      <span
        className="book__property__value"
      >
        mockGenre
         
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Publish Date: 
      </span>
      <span
        className="book__property__value"
      >
        02/02/2019
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Author Name: 
      </span>
      <span
        className="book__property__value"
      >
        mockName
         
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Author Gender: 
      </span>
      <span
        className="book__property__value"
      >
        mockGenre
         
      </span>
    </div>
  </div>
  <div
    className="book"
  >
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Name: 
      </span>
      <span
        className="book__property__value"
      >
        mockName1
         
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Genre: 
      </span>
      <span
        className="book__property__value"
      >
        mockGenre1
         
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Publish Date: 
      </span>
      <span
        className="book__property__value"
      >
        02/02/2019
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Author Name: 
      </span>
      <span
        className="book__property__value"
      >
        mockName1
         
      </span>
    </div>
    <div
      className="book__property"
    >
      <span
        className="book__property__name"
      >
        Author Gender: 
      </span>
      <span
        className="book__property__value"
      >
        mockGenre1
         
      </span>
    </div>
  </div>
</div>
`;
